connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java

```diff
@@ -0,0 +1,537 @@
+        deleted.connectors().forEach(c ->
+                toRevoke.computeIfAbsent(
+                    connectorOwners.get(c),
+                    v -> ConnectorsAndTasks.embed(new ArrayList<>(), new ArrayList<>()))
+                    .connectors().add(c));
+
+        // Add the tasks that have been deleted to the revoked set
+        deleted.tasks().forEach(t ->
+                toRevoke.computeIfAbsent(
+                    taskOwners.get(t),
+                    v -> ConnectorsAndTasks.embed(new ArrayList<>(), new ArrayList<>()))
+                    .tasks().add(t));
+        log.debug("Connectors and tasks to revoke: {}", toRevoke);
+
+        // Recompute the complete assignment excluding the deleted connectors-and-tasks
+        completeWorkerAssignment = workerAssignment(memberConfigs, deleted);
+        connectorAssignments =
+                completeWorkerAssignment.stream().collect(Collectors.toMap(WorkerLoad::worker, WorkerLoad::connectors));
+        taskAssignments =
+                completeWorkerAssignment.stream().collect(Collectors.toMap(WorkerLoad::worker, WorkerLoad::tasks));
+
+        if (!lostAssignments.isEmpty()) {
+            handleLostAssignments(lostAssignments, newSubmissions, completeWorkerAssignment);
+        }
+
+        assignConnectors(completeWorkerAssignment, newSubmissions.connectors());
+        assignTasks(completeWorkerAssignment, newSubmissions.tasks());
+
+        log.debug("Complete assignments: {}", currentWorkerAssignment);
+        log.debug("New complete assignments: {}", completeWorkerAssignment);
+
+        Map<String, Collection<String>> currentConnectorAssignments =
+                currentWorkerAssignment.stream().collect(Collectors.toMap(WorkerLoad::worker, WorkerLoad::connectors));
+
+        Map<String, Collection<ConnectorTaskId>> currentTaskAssignments =
+                currentWorkerAssignment.stream().collect(Collectors.toMap(WorkerLoad::worker, WorkerLoad::tasks));
+
+        Map<String, Collection<String>> incrementalConnectorAssignments =
+                diff(connectorAssignments, currentConnectorAssignments);
+
+        Map<String, Collection<ConnectorTaskId>> incrementalTaskAssignments =
+                diff(taskAssignments, currentTaskAssignments);
+
+        log.debug("Incremental Connector assignments: {}", incrementalConnectorAssignments);
+        log.debug("Incremental Task assignments: {}", incrementalTaskAssignments);
+
+        coordinator.leaderState(new LeaderState(memberConfigs, connectorAssignments, taskAssignments));
+
+        Map<String, ConnectAssignment> assignments =
+                fillAssignments(memberConfigs.keySet(), Assignment.NO_ERROR, leaderId,
+                                memberConfigs.get(leaderId).url(), maxOffset, incrementalConnectorAssignments,
+                                incrementalTaskAssignments, toRevoke, delay);
+
+        previousAssignment = ConnectorsAndTasks.embed(
```

Comment:
I was wondering about the thread safety of this assignment, but I see that calls to this method are protected by upstream synchronization in AbstractCoordinator. Maybe we should include a note about thread safety in the class javadoc?

Reply:
Added a note at the class level for both `EagerAssignor` and `IncrementalCooperativeAssignor`. We could potentially add a more detailed note such as the one in `AbstractCoordinator` on locking.